Add fast sync algorithm.

shamil-gadelshin opened the pull request:
This PR adds the last part of the fast-sync algorithm. The algorithm described in the [related spec](https://github.com/subspace/protocol-specs/blob/main/docs/consensus/consensus_chain.md#fast-sync) and in the comment for `FastSyncer` struct. The PR contains the new algorithm, its integration within the existing DSN sync, and CLI and config parameters changes required to enable the algorithm. Although to the best of my knowledge, the algorithm works as expected it should not be used until we develop a related solution for fraud proofs otherwise fast-sync will fail because of the missing state after reaching the fraud-proof.

### CLI and config changes
- new `fast-sync` parameter
- starting with `SyncMode::LightState` when `fast-sync` is enabled
- modification of the `state-pruning` parameters to support N blocks pruning (required by `SyncMode::LightState`)


### Notable changes for fast sync integration
- PoT check update to support fast-sync
- introducing `block-gap` clean on start when fast-sync enabled (Substrate detects it and inserts it repeatedly)
- `Reconstructor` is passed to DSN sync from the fast-sync algorithm to preserve the correct inner state

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

shamil-gadelshin committed:
node: Add fast-sync algorithm.

shamil-gadelshin committed:
node: Modify CLI configuration to support fast-sync.

shamil-gadelshin committed:
Add fast-sync engine.

shamil-gadelshin committed:
Add raw block importer.

shamil-gadelshin committed:
Integrate fast-sync algorithm into dsn-sync

shamil-gadelshin committed:
Simplify fast-sync-engine.

nazar-pc commented on crates/subspace-service/src/sync_from_dsn/fast_sync.rs:
```diff
@@ -0,0 +1,587 @@
+    /// - 2. add the last block the second last segment (as raw - without checks and execution),
+    /// - 3. download the state for the first block of the last segment,
+    ///     - add the first block of the last segment as raw,
+    ///     - download state for it,
```
These steps do not correspond to the latest version of the specification

shamil-gadelshin replied:
Will answer below to a similar comment.

nazar-pc commented on crates/subspace-service/src/sync_from_dsn/fast_sync.rs:
```diff
@@ -0,0 +1,587 @@
+            .await
+        {
+            error!(?error, "Failed to download segment headers.");
+            return Err(error);
```
I'm not sure we need to print errors if we also return them. Wrapping them with context would be helpful though.

nazar-pc commented on crates/subspace-service/src/sync_from_dsn/fast_sync.rs:
```diff
@@ -0,0 +1,587 @@
+        let last_segment_header = self
+            .segment_headers_store
+            .get_segment_header(last_segment_index)
+            .expect("We get segment index from the same storage. It should be present.");
```
This should look more like a proof:
```suggestion
            .expect("Last segment index exists; qed");
```

It is obvious that last segment index returned exists or else it wouldn't be returned as the last once, which is exactly what we needed to prove. Using "shoud" sounds like a wish, not a hard undebatable fact.

nazar-pc commented on crates/subspace-service/src/sync_from_dsn/fast_sync.rs:
```diff
@@ -0,0 +1,587 @@
+            .get_segment_header(last_segment_index)
+            .expect("We get segment index from the same storage. It should be present.");
+
+        let second_last_segment_index = last_segment_header.segment_index() - 1.into();
```
```suggestion
        let second_last_segment_index = last_segment_header.segment_index() - SegmentIndex::ONE;
```

nazar-pc commented on crates/subspace-service/src/sync_from_dsn/fast_sync.rs:
```diff
@@ -0,0 +1,587 @@
+            self.piece_getter,
+            &mut reconstructor,
+        )
+        .await?;
```
This is not how algorithm should work according to the spec though. You're supposed to check the segment header and only download second last segment header in case it is actually necessary, which is most of the time, but not always.

shamil-gadelshin replied:
The spec version will likely work, however, on the implementation level I have doubts - in the absolute majority of cases we will go into the coded branch, do we need this special case that complicates the code and requires support but will have a negligible effect? And more importantly, we don't have the testing infrastructure to verify the code, so, the spec theoretic version IMO has more drawbacks than benefits.

nazar-pc replied:
The only thing that needs to be done is an `if` branch. If the last block in second last segment was included partially then we download the segment and feed it into reconstructor, otherwise nothing needs to be downloaded or fed into reconstructor and we can proceed with the last segment only. This is a very simple check and I don't understand why we should intentionally not implement the spec and describe a different version in the function signature.

And such things are already tested in the archiver, you can easily logically confirm how it will behave without actually running the code.

I do not think I understand what the drawbacks there are in following the spec.

shamil-gadelshin replied:
The reason is that we won't test this code and it can fail even if seems logical which happened multiple times with other simple changes. It's likely possible to emulate this case when we have a complete last archived block. I checked the current blockchain - we don't have it currently. I added a TODO with a reference to the spec difference.

nazar-pc replied:
I'm sorry but it makes no sense to me. If we are so afraid to make changes we'll not make progress at any meaningful rate. Writing code that is intentionally wrong against the spec and even strictly speaking broken in general case makes even less sense to me.

nazar-pc commented on crates/subspace-service/src/sync_from_dsn/fast_sync.rs:
```diff
@@ -0,0 +1,587 @@
+            return Err(Error::Other(
+                "Unexpected block array length for the second last segment.".into(),
+            ));
+        }
```
I do not understand why any of this matters for sync process

shamil-gadelshin replied:
The error seems straightforward: there are indexing operations below. (I changed the condition though).

nazar-pc commented on crates/subspace-node/src/commands/run/consensus.rs:
```diff
@@ -397,6 +404,10 @@ pub(super) struct ConsensusChainOptions {
+
+    /// Experimental support of state-only sync using DSN.
+    #[arg(long, default_value_t = false)]
+    fast_sync: bool,
```
I don't think this is the right API, it doesn't scale.

Substrate CLI had this, which we have removed since we only supported full sync:
```
      --sync <SYNC_MODE>
          Blockchain syncing mode.
          
          [default: full]

          Possible values:
          - full:        Full sync. Download end verify all blocks
          - fast:        Download blocks without executing them. Download latest state with proofs
          - fast-unsafe: Download blocks without executing them. Download latest state without proofs
          - warp:        Prove finality and download the latest state
```

I think we should bring `--sync` back and this particular mode can be called "fast" (even though it will be a bit confusing for those farmiliar with Substrate). I'm also not sure it is a good idea to have state pruning defined by a number since IIRC it will prune state regardless of finality, which means node will likely fail to restart in some cases due to state being unavailable at the point of archiver initialization, which is problematic.

This is why I mentioned in the past that we should probably call it "farmer sync" and not support defining state pruning mode explicitly, we should manually prune things when we need to, but not sooner (hence https://github.com/paritytech/polkadot-sdk/issues/1570 from September of last year).

So for now I'd vote for `--sync farmer` and do not allow user to specify `--blocks-pruning` or `--state-pruning` when these are used. Not sure what to do about state pruning yet though. Did you confirm that it doesn't remove non-finalized state?

shamil-gadelshin replied:
Makes sense. I will add this later either within this PR or in a separate one.

nazar-pc commented on crates/subspace-node/src/commands/run/consensus.rs:
```diff
@@ -584,7 +596,13 @@ pub(super) fn create_consensus_chain_configuration(
-    let consensus_chain_config = Configuration::from(consensus_chain_config);
+    let mut consensus_chain_config = Configuration::from(consensus_chain_config);
+    if fast_sync {
+        consensus_chain_config.network.sync_mode = SyncMode::LightState {
```
There do not seem to be any TODOs that this is a hack that should go away once we don't have to do this anymore as described in https://github.com/paritytech/polkadot-sdk/issues/4407

nazar-pc commented on crates/subspace-service/src/sync_from_dsn/import_blocks.rs:
```diff
@@ -136,7 +136,7 @@ where
             .expect("Statically guaranteed to exist, see checks above; qed");
 
-        trace!(
+        debug!(
```
This is a trace message to me. We already have a debug message about segment index being processed.

nazar-pc commented on crates/subspace-service/src/lib.rs:
```diff
@@ -318,6 +321,22 @@ where
+                                return true;
+                            }
+                        }
+                    }
```
Why are we skipping PoT verification here? PR description "PoT check update to support fast-sync" is not sufficient to understand why this is required and I do not remember discussing this.

shamil-gadelshin replied:
It's yet another special case for fast-sync. Did you mean something else?

nazar-pc replied:
I can guess from the code that this is some kind of special case, but I don't understand what that special case is, what issues it causes and why you decided to fix it this way out of a set of possibilities. I sometimes can infer those things, but you'll have to help me understand it in this case.

shamil-gadelshin replied:
AFAIR PoT verification requires runtime parameters check and it will fail without the existing state. This snippet disables the check that will fail for this reason for blocks before the fast-sync target block.

nazar-pc replied:
But we always import blocks with state, I'm not sure I'm following what you're saying.

nazar-pc replied:
So the reason you had to add it here is the side effect of incorrectly importing two raw blocks instead of one (the first one being without state). This is also the reason a more advanced version of https://github.com/subspace/subspace/pull/2767 failed: there existed block without state that should not have been there. The fact that that extra block is present will also cause archiver to fail to restart if interrupted right after insertion of the first block into database.

Basically insertion of two raw blocks is incorrect and should be fixed.

Once fixed, the failure will happen at the beginning of the function when attempting to read parent header and should be handled there accordingly.

I checked the code and I believe we should never ever have invalid votes included in the block and parent hash always comes from runtime's state and not user input, thus it should be safe to return `true` from this function in that special case.

nazar-pc replied:
Fixed in https://github.com/subspace/subspace/pull/2779

nazar-pc commented on crates/subspace-service/src/lib.rs:
```diff
@@ -490,17 +511,21 @@ where
 
     let executor = Arc::new(executor);
 
+    let fast_sync_state: Arc<Mutex<Option<NumberFor<Block>>>> = Default::default();
```
This is usually more readable and a tiny bit shorter:
```suggestion
    let fast_sync_state = Arc::<Mutex<Option<NumberFor<Block>>>>::default();
```

nazar-pc commented on crates/subspace-service/src/lib.rs:
```diff
@@ -709,8 +735,19 @@ where
+            debug!(client_info=?client.info(), "Clear block gap after fast-sync.");
+            client.clear_block_gap();
+        }
+    }
```
This feels like a completely random place to clear gap sync. Can it be moved right next to the code that triggers gap sync in the first place?

shamil-gadelshin replied:
We need this operation before the regular sync starts. What place do you advise?

nazar-pc replied:
Is this suggestion viable?:
> Can it be moved right next to the code that triggers gap sync in the first place?

shamil-gadelshin replied:
I prefer to clear the gap before driving any networking futures. This is why it's placed in the initialization code far away from the sync-code.

nazar-pc replied:
I don't think I understand the rationate here. If gap sync is triggered by some specific behavior then we should clear gap sync after we trigger that behavior. If the behavior is importing blocks in specific order then you will absolutely drive network prior to gap sync because that is how you end up with block you can physically import in the first place.

nazar-pc commented on crates/subspace-service/src/lib.rs:
```diff
@@ -806,7 +843,7 @@ where
     };
 
-    let import_queue_service = import_queue.service();
+    let import_queue_service = Arc::new(tokio::sync::Mutex::new(import_queue.service()));
```
Do you just need import queue service in multiple places? If so you can call `import_queue.service()` more than once and get two service instances with handles to the same import queue. Or is there a different reason for this?

shamil-gadelshin replied:
At some point, there were three instances passed across several methods that seemed awkward. I managed to work with two instances but I left shared ref anyway.

nazar-pc replied:
You're still sharing it this way, what is the benefit of `Arc<Mutex<T>>` comparing to just `T` here? Seems like unnecessary complexity to me.

shamil-gadelshin replied:
Am I correct that you prefer passing two instances of service through methods instead?

nazar-pc replied:
Absolutely, it is like having a Arc and instead of accessing it from two places directly you wrapped it in additional `Arc<Mutex>>`. I simply do not understand why would you even do that.

nazar-pc commented on crates/subspace-service/src/sync_from_dsn.rs:
```diff
@@ -47,49 +53,72 @@ enum NotificationReason {
-            client.as_ref(),
-            import_queue_service.as_mut(),
+            client.clone(),
+            import_queue_service,
```
Why owned values? I think references should still work just fine.

shamil-gadelshin replied:
They are both `Arc` references

nazar-pc replied:
I can see that, but the question above remains

shamil-gadelshin replied:
There are type requirements down the stack.

nazar-pc replied:
Which types? They were not required before, what has changed in a way that makes it mandatory?

nazar-pc commented on crates/subspace-service/src/sync_from_dsn.rs:
```diff
@@ -251,19 +290,70 @@ where
+                    sc_service::Error::Other("Can't send sync notification reason.".into())
+                })?;
+        }
+    }
```
As discussed before this doesn't seem to belong to full sync from DSN, right now it is a one-off separate sync that happens before DSN sync . DSN sync is not a continuation of this sync because we will have new enough state for regular sync to kick in and take it from there, no need to pass reconstructor through everything.

shamil-gadelshin replied:
I added the shared reconstructor when I encountered an error. Why do you think there is no need to pass the reconstructor? I appreciate more details here. There are two more reasons to place it here: fast-sync IS DSN-sync, shared sync pause management. However, I'm open to suggestions here.

nazar-pc replied:
> I added the shared reconstructor when I encountered an error

You know the next natural question I'll have would be "which error did you encounter exactly"?

> Why do you think there is no need to pass the reconstructor? I appreciate more details here.

Everything worked fine before, so it is not on me to prove that with new code added this is not necessary, you should convince me that something that didn't exist before is now necessary (which is very possible, I just don't see it right now).

> There are two more reasons to place it here: fast-sync IS DSN-sync, shared sync pause management. However, I'm open to suggestions here.

Arguably it is not. Fast sync's primary goal is to import a single block from DSN: the very first block of the last segment. Even if node shuts down right after and DSN sync fails after restart, we can still continue regular Substrate sync successfully. The fact that we import other blocks from the last segment is an optimization (because we already have it, no need to throw them away), not a requirement.

And I already mentioned in one of the prior discussions that if we download source pieces of last segment sequentially we can both continue downloading last segment and import blocks concurrently before we've got the whole segment. It is important to distinguish what is the goal and the core feature we are trying to build and what are "nice to have" on top.

shamil-gadelshin replied:
> You know the next natural question I'll have would be "which error did you encounter exactly"?

It's hard to reproduce it on the fly but AFAIR the new reconstructor will result in a missing block on the edges of two segments (the last imported from the fast sync and the remaining segment obtained through DSN sync).

nazar-pc replied:
DSN sync is designed to resume from any point, it'll just have to potentially download some of the data twice, but as articulated a few times already we don't need to call DSN sync after fast sync in the first place because regular Substrate sync can pick up from there already.

nazar-pc commented on crates/subspace-service/src/sync_from_dsn.rs:
```diff
@@ -251,19 +290,70 @@ where
+        }
+    }
+
+    while let Some(reason) = notifications.next().await {
```
You have removed sync pausing when notification is received, this is incorrect and will cause issues

shamil-gadelshin replied:
I restored the pause. Please, share possible issues.

nazar-pc commented on crates/subspace-service/src/sync_from_dsn.rs:
```diff
@@ -212,30 +241,40 @@ async fn create_substrate_network_observer<Block>(
+    sync_service: Arc<SyncingService<Block>>,
 ) -> Result<(), sc_service::Error>
 where
+    B: sc_client_api::Backend<Block>,
```
If `B` is first here, it should logically be first in the list of function generics as well, it is the case in all other cases here and elsewhere

nazar-pc reviewed (commented):
Left a bunch of comments and questions, will have to re-review again afterwards. Sorry for delay.

shamil-gadelshin committed:
Apply review suggestions.

- refactor comments, error messages, expect messages
- restore pause for DSN-sync
- add comments
- other minor changes

nazar-pc commented on crates/subspace-service/src/sync_from_dsn/fast_sync.rs:
```diff
@@ -0,0 +1,587 @@
+            return Err(Error::Other(
+                "Unexpected block array length for the second last segment.".into(),
+            ));
+        }
```
I think I understand now what you meant to do here, it is just code that doesn't actually talk about it and doesn't check the actual invariant you care about.

The only thing (an nothing else) that we actually care about in fast sync is to import the first block of the last archived segment, this is why spec talks about that and not other things. There is no requirement for 2 blocks to be present, especially not in second last segment.

The edge-case though might be that we could hypothetically have (not practically, but it is still worth supporting IMHO) a block larger that a segment. In that case we may be unable to reconstruct the first block of last segment even if we download two last segments.

So what the code should actually is very close to what spec says:
* once segment headers are known check the header of second last segment and see if last block there was archived or partially
* if fully we only need to download the last segment and nothing else
* if we have a partial block though then we need to check if third last segment ended with the same partial block number because if it did then we need to keep looking forward to the beginning of that block such that we can get the full block of the last segment header regardless of where it starts

With above description it should be clear that this check doesn't make sense. For example it will fail in the perfectly fine situation where second last segment contains end of one block and beginning of another (first block of last segment). In this case we have 0 blocks reconstructed from second last segment alone, yet we are able to successfully reconstruct first block of last segment that the specification actually requires.

shamil-gadelshin replied:
Both checks were introduced because of the indexing operations (to fail with the understandable error). However, as you pointed out we don't support this case - I will remove it instead considering indexing infallible in this case.

nazar-pc commented on crates/subspace-service/src/sync_from_dsn/fast_sync.rs:
```diff
@@ -0,0 +1,587 @@
+            return Err(Error::Other(
+                "Unexpected block array length for the last segment.".into(),
+            ));
+        }
```
This check is unnecessary, the only thing we need is a single block. Other blocks can be imported even via regular Substrate sync without issues.

Here is the invariant from the spec you should check instead:
```suggestion
        if blocks.is_empty() {
            return Err(Error::Other(
                "Failed to reconstruct first block of the last segment".into(),
            ));
        }
```

shamil-gadelshin replied:
Both checks were introduced because of the indexing operations (to fail with the understandable error). However, as you pointed out we don't support this case - I will remove it instead considering indexing infallible.

shamil-gadelshin committed:
Change peer source for fast-sync.

shamil-gadelshin committed:
Update fast sync method.

nazar-pc commented on crates/subspace-service/src/sync_from_dsn/fast_sync.rs:
```diff
@@ -507,14 +507,18 @@ where
+                    .peers_info()
                     .await
-                    .expect("Network service must be available.");
+                    .expect("Network service must be available.")
```
Is there a local invariant guarantees that? I'd return an error instead if not, it is not obvious why it is guaranteed here.

shamil-gadelshin committed:
Refactor fast-sync.

nazar-pc commented on crates/subspace-service/src/sync_from_dsn/fast_sync.rs:
```diff
@@ -0,0 +1,581 @@
+                current_block_number,
+                BlockOrigin::NetworkInitialSync,
+            )
+            .await?;
```
You should schedule all the blocks together rather than one by one.

Scheduling them one by one means issues with verification concurrency and huge amount of logs with errors for every single block in case things go wrong. Check how DSN sync implementation schedules block import.

UPD: Fixed in https://github.com/subspace/subspace/pull/2779

nazar-pc committed:
Only sync state from peers that have a higher best block number than block number of the state we're trying to download

nazar-pc committed:
Import lists of blocks together instead of individually to leverage verification concurrency and better error handling

nazar-pc commented on crates/subspace-service/src/sync_from_dsn/fast_sync.rs:
```diff
@@ -0,0 +1,581 @@
+
+        let blocks_in_second_last_segment = second_last_segment_blocks.len();
+        debug!(
+            "Second last segment blocks downloaded (SegmentId={}): {}-{}",
```
There is no `SegmentId` in the protocol, there are just segment indices. I'd use `%second_last_segment_index` instead of this. On a similar note I'd use `{}..={}` instead of `{}-{}` for block numbers since we're in Rust.

nazar-pc committed:
Remove incorrect of last block from second last segment and add TODOs

nazar-pc reviewed (commented):
Fixed a few key issues in https://github.com/subspace/subspace/pull/2779 except non-atomic state import and didn't implement support for blocks that cover multiple segments because code will have to be refactored anyway and it can be done then.

https://github.com/subspace/subspace/pull/2779 needs to be merged into this PR, then further adjustments and cleanups need to be added, after which we'll squash numerous commits into one, maybe a couple, with clear additions according to spec and minimal changes elsewhere.

shamil-gadelshin committed:
Merge pull request #2779 from subspace/add-fast-sync-algorithm-fixes

Add fast sync algorithm fixes

shamil-gadelshin committed:
Move fast-sync out of DSN-sync scope.

shamil-gadelshin committed:
Modify sync CLI arguments.

nazar-pc committed:
Inline both `import_raw_block` and `create_raw_block` into the place where they are used

shamil-gadelshin committed:
Merge branch 'main' into add-fast-sync-algorithm

# Conflicts:
#	crates/subspace-service/Cargo.toml

shamil-gadelshin committed:
Remove peer-management elements from fast-sync.

shamil-gadelshin committed:
Merge pull request #2785 from subspace/inline-import-raw-block

Inline both `import_raw_block` and `create_raw_block` into the place where they are used

shamil-gadelshin committed:
Add some comments.

nazar-pc commented on crates/subspace-service/src/sync_from_dsn/fast_sync.rs:
```diff
@@ -260,6 +263,7 @@ where
         );
 
         let mut import_block = BlockImportParams::new(
+            // We support the case when the block with extra-big blocks.
```
Hm... this comment is in various places but I don't think it actually explains anything.

Here specifically the reason we have this check is that the block we are importing is the last one, there are no other blocks to import afterwards, hence we need to notify Substrate sync with `BlockOrigin::NetworkBroadcast` instead of `BlockOrigin::NetworkInitialSync`.

I don't see how `We support the case when the block with extra-big blocks.` would help anyone understand that, I don't see a direct connection at all.

nazar-pc committed:
Revert unnecessary changes to DSN sync code

nazar-pc commented on crates/subspace-service/src/lib.rs:
```diff
@@ -709,6 +722,16 @@ where
+            debug!(client_info=?client.info(), "Clear block gap after fast-sync.");
+            client.clear_block_gap();
+        }
+    }
```
IIRC this should have been removed, right?

nazar-pc committed:
Remove `FastSyncResult` that is no longer necessary

nazar-pc committed:
Remove unnecessary retrieval of last segment header

nazar-pc committed:
Rename `download_segment_headers` to `sync_segment_headers` and simplify its API

nazar-pc committed:
Make fast sync not assume block size

nazar-pc committed:
Make fast sync import block atomically, simplify code

nazar-pc committed:
Update Subspace block import to be better suited for custom sync mechanisms

nazar-pc committed:
Small corrections and cleanups

nazar-pc committed:
Remove unnecessary `Mutex` and unused generics

nazar-pc committed:
Simplify fast sync engine API, add support for `fork_id`

nazar-pc committed:
Simplify networking in fast sync engine, simplify generics further

nazar-pc committed:
Correct check for fast sync start and fix sync mode resetting after (optional) fast sync

nazar-pc committed:
Simplify block import without resorting to `apply_block` low-level API

shamil-gadelshin committed:
Toggle skip_proof variable for state downloading.

shamil-gadelshin committed:
Merge pull request #2795 from subspace/revert-dsn-sync-changes

Revert unnecessary changes to DSN sync code

nazar-pc committed:
Merge pull request #2796 from subspace/fast-sync-improvements

Fast sync improvements

nazar-pc committed:
Merge pull request #2797 from subspace/atomic-fast-sync-block-import

Atomic fast sync block import

nazar-pc committed:
Merge pull request #2798 from subspace/fast-sync-cleanup

Fast sync cleanup

nazar-pc committed:
Do not finalize blocks that do not exist

nazar-pc committed:
Fix condition for clearing block gap

nazar-pc committed:
Print info after fast sync

nazar-pc committed:
Merge pull request #2799 from subspace/simplify-fast-sync

Simplify fast sync

shamil-gadelshin committed:
Update fast-sync after latest polkadot-sdk

shamil-gadelshin committed:
Merge branch 'main' into add-fast-sync-algorithm

shamil-gadelshin committed:
Rename fast-sync -> snap-sync

shamil-gadelshin committed:
Merge branch 'main' into add-fast-sync-algorithm

# Conflicts:
#	crates/subspace-service/Cargo.toml

nazar-pc commented on crates/subspace-node/src/commands/run/consensus.rs:
```diff
@@ -171,17 +173,21 @@ impl FromStr for StatePruningMode {
-            _ => Err("Invalid state pruning mode specified".to_string()),
+            n => n
+                .parse()
+                .map_err(|_| "Invalid block pruning mode specified".to_string())
```
```suggestion
                .map_err(|_| "Invalid block state mode specified".to_string())
```

nazar-pc commented on crates/subspace-node/src/commands/run/consensus.rs:
```diff
@@ -387,16 +394,60 @@ pub(super) struct ConsensusChainOptions {
+    /// Full sync. Download and verify all blocks.
+    Full,
+    /// Download blocks from DSN.
+    Dsn,
```
I'm not sure why we have these as separate modes, nothing in the code seems to make any distinction between them. Also we don't need separate `ConfigSyncMode` and `ChainSyncMode`, just `ChainSyncMode` that implements `FromStr` is sufficient.

shamil-gadelshin replied:
We use `Snap` and `not Full`, so we use `Dsn` implicitly. Two types `ChainSyncMode` and `ConfigSyncMode` were introduced to avoid adding `clap` dependency into `subspace-service`.

nazar-pc replied:
I didn't understand what you wanted to say with the first sentence, can you rephrase?

As for Clap, you don't need it. Enum only needs to implement `FromStr` trait IIRC.

shamil-gadelshin replied:
We use `Dsn` value implicitly without a code reference as following: we either use only `Full` (regular) sync or DSN sync option when we can run `Snap` sync first.

> As for Clap, you don't need it. Enum only needs to implement FromStr trait IIRC.

Indeed!

nazar-pc replied:
> We use Dsn value implicitly without a code reference as following: we either use only Full (regular) sync or DSN sync option when we can run Snap sync first.

I'm sorry, it still makes no sense to me, at least it doesn't seem to answer the question I had originally.

shamil-gadelshin replied:
Previously, we had two sync modes - regular and DSN, the latter was enabled by `--sync-from-dsn` and started a dedicated task.
Now, we have three modes: full, dsn, and snap. Snap starts in the same dedicated task as Dsn sync. 
We start the dedicated task by checking `if !config.sync.is_full()`. "Not Full" includes Dsn and Snap sync modes. Snap sync can be activated by `Snap` sync mode before the Dsn worker in the dedicated sync task. 
Thus we use Dsn sync mode implicitly without referencing it in the code. 

I hope it makes more sense now. Feel free to ask additional questions.

nazar-pc replied:
I missed `is_full` check somehow. You made `Full` sync the default, which will disable DSN sync if I understand correctly. I'd say we should simply remove ability to disable sync from DSN. There should be just `Full` and `Snap`. No need to have `Dsn` as a third option, users will not be able to sync fully without DSN anyway (almost all nodes are pruned).

shamil-gadelshin replied:
Will domain operators be affected by this change? We're changing our previous semantics.

nazar-pc replied:
What semantics? DSN sync was enabled before unless it was disabled explicitly with CLI option, so it was opt-out. Now you made it opt-in instead and I suggest to return to previous behavior and remove option that disabled sync from DSN so it is always enabled.

nazar-pc commented on crates/subspace-service/src/sync_from_dsn/snap_sync_engine.rs:
```diff
@@ -0,0 +1,224 @@
+#![allow(dead_code)]
```
Time to remove this

nazar-pc commented on crates/subspace-service/src/lib.rs:
```diff
@@ -924,6 +953,11 @@ where
                 Box::pin(async move {
+                    // Run snap-sync before DSN-sync.
+                    if config.sync == ChainSyncMode::Snap {
+                        let _ = snap_sync_task.await;
```
This will ensure that we get compiler warning if/when we decide to return something like `Result` out of it instead of current `()`:
```suggestion
                        snap_sync_task.await;
```

nazar-pc reviewed (commented):
Left some nits, this will then need to be squashed as we discussed previously

shamil-gadelshin committed:
Refactor fast-sync.

shamil-gadelshin committed:
Remove redundant `ConfigSyncMode` type

shamil-gadelshin committed:
Remove substrate sync option.

nazar-pc reviewed (commented):
I think this is good to go, I'll open squashed PR instead of 50 individual commits. Also CI has failed, looks like we have a flaky test in networking.
